File: src/main/java/tech/simter/util/BeanUtils.java
package tech.simter.util;

import org.modelmapper.ModelMapper;

/**
 * Bean property copy util tools.
 *
 * @author RJ
 */
public class BeanUtils {
  private static ModelMapper mapper;

  static {
    // initial {@link ModelMapper} instance
    mapper = new ModelMapper();
    mapper.getConfiguration().setFieldMatchingEnabled(true);
  }

  /**
   * Get the default {@link ModelMapper} instance,
   * <p>
   * <code>mapper.getConfiguration().setFieldMatchingEnabled(true)</code>
   *
   * @return {@link ModelMapper} instance
   */
  public static ModelMapper getDefaultMapper() {
    return mapper;
  }

  /**
   * Maps {@code sources} to an instance of {@code targetType}.
   *
   * @param targetType targetType type
   * @param sources    objects to copy from
   * @param <T>        type to copy to
   * @return fully mapped instance of {@code targetType}
   * @see ModelMapper#map(Object, Class) and ModelMapper#copy(Object, Object)
   */
  public static <T> T assign(Class<T> targetType, Object... sources) {
    if (sources == null || sources.length == 0) return null;

    T target = null;
    for (int i = 0; i < sources.length; i++) {
      if (i == 0) target = mapper.map(sources[i], targetType);
      else mapper.map(sources[i], target);
    }
    return target;
  }

  /**
   * Maps {@code sources} to a target instance.
   *
   * @param target  target instance
   * @param sources objects to copy from
   * @see ModelMapper#map(Object, Object)
   */
  public static void assign(Object target, Object... sources) {
    if (target == null || sources == null || sources.length == 0) return;
    for (Object source : sources) mapper.map(source, target);
  }
}

File: src/main/kotlin/tech/simter/util/BeanUtils.kt
package tech.simter.util

import org.modelmapper.ModelMapper
import org.modelmapper.Provider
import java.util.*
import kotlin.reflect.full.memberProperties

/**
 * Bean property copy util tools.
 *
 * @author RJ
 */
object BeanUtils {
  /**
   * Get the default [ModelMapper] instance,
   *
   *
   * `mapper.getConfiguration().setFieldMatchingEnabled(true)`
   *
   * @return [ModelMapper] instance
   */
  @Suppress("MemberVisibilityCanBePrivate")
  val defaultMapper: ModelMapper by lazy {
    // initial a ModelMapper instance
    val modelMapper = ModelMapper()
    modelMapper.configuration
      .setFieldMatchingEnabled(true)
      // Construct Set property to LinkedHashedSet instance to keep orders.
      .provider = Provider<Set<Any>> {
      if (it.requestedType.isAssignableFrom(Set::class.java)) LinkedHashSet() else null
    }
    modelMapper
  }

  /**
   * Maps `sources` to an instance of `targetType`.
   *
   * @param targetType targetType type
   * @param sources    objects to copy from
   * @param <T>        type to copy to
   * @return fully mapped instance of `targetType`
   * @see ModelMapper.map
   */
  fun <T> assign(targetType: Class<T>, vararg sources: Any): T? {
    if (sources.isEmpty()) return null

    var target: T? = null
    for (i in sources.indices) {
      if (i == 0)
        target = defaultMapper.map(sources[i], targetType)
      else
        defaultMapper.map(sources[i], target)
    }
    return target
  }

  /**
   * Maps `sources` to a target instance.
   *
   * @param target  target instance
   * @param sources objects to copy from
   * @see ModelMapper.map
   */
  fun assign(target: Any?, vararg sources: Any) {
    if (target == null || sources.isEmpty()) return
    for (source in sources) defaultMapper.map(source, target)
  }

  /**
   * Compare [newObj] with [oldObj] to find all difference properties in [newObj].
   *
   * Properties in [newObj] treat as difference if not match the [equalizer].
   * Default [equalizer] is simple implemented as value equals, two null value is treated as equals.
   */
  fun diff(
    newObj: Any,
    oldObj: Any,
    equalizer: (propertyName: String, newValue: Any?, oldValue: Any?) -> Boolean = { _, newValue, oldValue ->
      newValue == oldValue
    },
    ignoreNullValue: Boolean = false
  ): List<Diff> {
    var oldValues = if (oldObj is Map<*, *>) oldObj.mapKeys { it.key.toString() }
    else oldObj.javaClass.kotlin.memberProperties
      .filter { it.visibility == kotlin.reflect.KVisibility.PUBLIC } // only public properties
      .associateBy({ it.name }, { it.get(oldObj) })
    if (ignoreNullValue) oldValues = oldValues.filter { it.value != null } // filter not null value

    var newValues = if (newObj is Map<*, *>) newObj.mapKeys { it.key.toString() }
    else newObj.javaClass.kotlin.memberProperties
      .filter { it.visibility == kotlin.reflect.KVisibility.PUBLIC } // only public properties
      .associateBy({ it.name }, { it.get(newObj) })
    if (ignoreNullValue) newValues = newValues.filter { it.value != null } // filter not null value

    return diff(newValues = newValues, oldValues = oldValues, equalizer = equalizer)
  }

  /**
   * Find all difference values from [newValues] compare to [oldValues] by the [equalizer] with same key.
   *
   * Default [equalizer] is simple implemented as value equals, two null value is treated as equals.
   */
  fun diff(
    newValues: Map<String, Any?>,
    oldValues: Map<String, Any?>,
    equalizer: (propertyName: String, newValue: Any?, oldValue: Any?) -> Boolean = { _, newValue, oldValue ->
      newValue == oldValue
    }
  ): List<Diff> {
    return newValues.filterNot { equalizer(it.key, it.value, oldValues[it.key]) }
      .map { Diff(name = it.key, newValue = it.value, oldValue = oldValues[it.key]) }
  }

  /** A new and old value holder */
  data class Diff(
    val name: String,
    val newValue: Any?,
    val oldValue: Any?
  )
}